opal/mca/common/ofi/common_ofi.c

```diff
@@ -445,63 +445,164 @@ static int check_provider_attr(struct fi_info *provider_info, struct fi_info *pr
- *
- *     Uses a pci device to find an ancestor that contains a cpuset, and
- *     determines if it intersects with the cpuset that the process is bound to.
- *     if the process is not bound, or if a cpuset is unavailable for whatever
- *     reason, returns false. Otherwise, returns the result of
- *     hwloc_cpuset_intersects()
- */
-static bool compare_cpusets(hwloc_topology_t topology, struct fi_pci_attr pci)
+static int calculate_distances(pmix_device_distance_t **distances,
+                               size_t *ndist)
 {
-    bool result = false;
-    int ret;
-    hwloc_bitmap_t proc_cpuset;
-    hwloc_obj_t obj = NULL;
+    int ret = 0;
+    size_t ninfo;
+    pmix_info_t *info;
+    pmix_cpuset_t cpuset;
+    pmix_topology_t *pmix_topo;
+    pmix_device_type_t type = PMIX_DEVTYPE_OPENFABRICS |
+      PMIX_DEVTYPE_NETWORK;
+
+    PMIX_CPUSET_CONSTRUCT(&cpuset);
+    ret = PMIx_Get_cpuset(&cpuset, PMIX_CPUBIND_THREAD);
+    if (PMIX_SUCCESS != ret) {
+        goto out;
+    }
+
+    /* load the PMIX topology */
+    PMIX_TOPOLOGY_CREATE(pmix_topo, 1);
+    ret = PMIx_Load_topology(pmix_topo);
+    if (PMIX_SUCCESS != ret) {
+        goto out;
+    }
+
+    ninfo = 1;
+    PMIX_INFO_CREATE(info, ninfo);
+    PMIX_INFO_LOAD(&info[0], PMIX_DEVICE_TYPE, &type, PMIX_DEVTYPE);
+    ret = PMIx_Compute_distances(pmix_topo, &cpuset, info, ninfo, distances,
+                                 ndist);
+    PMIX_INFO_FREE(info, ninfo);
+
+    PMIX_TOPOLOGY_FREE(pmix_topo, 1);
+out:
+    return ret;
+}
 
-    /* Cannot find topology info if no topology is found */
-    if (NULL == topology) {
-        return false;
+static pmix_device_distance_t *get_nearest_nics(pmix_value_t *val,
+                                                int *num_distances)
+{
+    size_t ndist;
+    pmix_topology_t *topo;
+    int ret, i, idx = 0;
+    pmix_data_array_t *dptr;
+    uint16_t near = USHRT_MAX;
+    pmix_info_t directive;
+    pmix_device_distance_t *distances, *nearest = NULL;
+
+    PMIX_INFO_LOAD(&directive, PMIX_OPTIONAL, NULL, PMIX_BOOL);
+    ret = PMIx_Get(&opal_process_info.myprocid,
+             PMIX_DEVICE_DISTANCES, &directive, 1, &val);
+    PMIX_INFO_DESTRUCT(&directive);
+    if (ret != PMIX_SUCCESS || !val) {
+        ret = calculate_distances(&distances, &ndist);
+        if (ret)
+            goto out;
+        goto find_nearest;
+    }
+
+    if (PMIX_DATA_ARRAY != val->type) {
+        goto out;
+    }
+    dptr = val->data.darray;
+    if (NULL == dptr) {
+        goto out;
+    }
+    if (PMIX_DEVICE_DIST != dptr->type) {
+        goto out;
+    }
+
+    distances = (pmix_device_distance_t*)dptr->array;
+    ndist = dptr->size;
+
+find_nearest:
+    nearest = calloc(sizeof(*distances), ndist);
+    if (!nearest)
+        goto out;
+
+    for (i = 0; i < ndist; i++) {
+        if (distances[i].mindist < near) {
+            idx = 0;
+            near = distances[i].mindist;
+            nearest[idx] = distances[i];
```

Comment:
I'm having trouble figuring out what the arrays represent. Can you add doxygen comments to the functions that explains them?

Reply:
Thanks a lot, the doxygen comments helped tremendously in understanding what's going on